File: admin-web/lib/ozellikler.ts
"use client";

import useSWR from "swr";

import { jsonFetcher } from "@/lib/fetcher";

/**
 * (P221) SUNUCUDAN GELEN OZELLIK BAYRAKLARI — web.
 *
 * =========================================================================
 * MOBILLE AYNI KURAL
 * =========================================================================
 * Bayrak SUNUCUDA; acmak icin yeni bir dagitim degil tek bir ortam
 * degiskeni yetiyor. VARSAYILAN KAPALI: ag hatasi, eski sunucu ya da
 * bozuk govde — hepsinde `false`.
 *
 * "Bilmiyorsak acalim" demek, hazir olmayan bir yuzeyi kullaniciya
 * gostermek olurdu.
 */
export interface OzellikBayraklari {
  dukkan: boolean;
}

export function useDukkanAcik(): boolean {
  const { data } = useSWR<OzellikBayraklari>("/api/ozellikler", jsonFetcher);
  // `data?.dukkan === true`: yukleme (undefined), hata (undefined) ve
  // beklenmedik tip — hepsi KAPALI.
  return data?.dukkan === true;
}

File: backend/app/routers/ozellikler.py
"""(P221) GENEL OZELLIK BAYRAKLARI — istemcinin acilista okudugu yuzey.

===========================================================================
NEDEN AYRI BIR UC
===========================================================================
Bir ozelligi acmak/kapatmak SUNUCUNUN karari olmali. Bayrak uygulamanin
icinde olsaydi, acmak icin YENI SURUM ve MAGAZA TURU gerekirdi — ve
magaza incelemesi gunler surer. Yayinlanmis bir sekmeyi ayni gun acmak
ancak sunucudan mumkun.

`POST /surum/kontrol`a EKLENMEDI: o uc "surumun guncel mi" sorusunu
yanitliyor ve yanitini bir POLITIKA TABLOSUNDAN aliyor. Ozellik
bayraklari ortam yapilandirmasindan geliyor; iki farkli kaynagi tek
yanitta toplamak, birinin degismesi digerinin onbellegini de bozmasi
demekti.

===========================================================================
KIMLIK GEREKTIRMEZ — VE BU BILINCLI
===========================================================================
Istemci bu ucu OTURUM ACMADAN, acilista cagiriyor: hangi sekmelerin
gorunecegi giris ekranindan once belli olmali. Yanit hicbir tesise ya da
kisiye ozel veri TASIMIYOR — yalniz "su ozellik acik mi".

===========================================================================
VARSAYILAN KAPALI
===========================================================================
Yanit uretilemezse (ag hatasi, uc yok, eski sunucu) istemci KAPALI
varsayar. Bu, ozelligin sessizce acilmasindan iyidir: hazir olmayan bir
yuzeyi kullaniciya gostermektense gostermemek yeglenir.
"""
from __future__ import annotations

from fastapi import APIRouter

from ..config import settings
from ..schemas import OzellikBayraklari

router = APIRouter(tags=["ozellikler"])


@router.get("/ozellikler", response_model=OzellikBayraklari)
async def ozellikler() -> OzellikBayraklari:
    """Istemcinin acilista okudugu ozellik bayraklari.

    Doner: `{"dukkan": bool}`

    YENI BAYRAK EKLERKEN: alani `OzellikBayraklari`ya VARSAYILANIYLA
    ekleyin. Eski istemciler bilmedikleri alani yok sayar; yeni
    istemciler eski sunucudan alani almadiginda kendi varsayilanini
    kullanir. Iki yonde de kirilma olmaz.
    """
    return OzellikBayraklari(dukkan=settings.dukkan_mobil_acik)
